File: rag/app/tag.py
#
#  Copyright 2025 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
import json
import re
import csv
from copy import deepcopy

from deepdoc.parser.utils import get_text
from rag.app.qa import Excel
from rag.nlp import rag_tokenizer
from common import settings


def beAdoc(d, q, a, eng, row_num=-1):
    d["content_with_weight"] = q
    d["content_ltks"] = rag_tokenizer.tokenize(q)
    d["content_sm_ltks"] = rag_tokenizer.fine_grained_tokenize(d["content_ltks"])
    d["tag_kwd"] = [t.strip().replace(".", "_") for t in a.split(",") if t.strip()]
    if row_num >= 0:
        d["top_int"] = [row_num]
    return d


def chunk(filename, binary=None, lang="Chinese", callback=None, **kwargs):
    """
        Excel and csv(txt) format files are supported.
        If the file is in excel format, there should be 2 column content and tags without header.
        And content column is ahead of tags column.
        And it's O.K if it has multiple sheets as long as the columns are rightly composed.

        If it's in csv format, it should be UTF-8 encoded. Use TAB as delimiter to separate content and tags.

        All the deformed lines will be ignored.
        Every pair will be treated as a chunk.
    """
    eng = lang.lower() == "english"
    res = []
    doc = {
        "docnm_kwd": filename,
        "title_tks": rag_tokenizer.tokenize(re.sub(r"\.[a-zA-Z]+$", "", filename))
    }
    if re.search(r"\.xlsx?$", filename, re.IGNORECASE):
        callback(0.1, "Start to parse.")
        excel_parser = Excel()
        for ii, (q, a) in enumerate(excel_parser(filename, binary, callback)):
            res.append(beAdoc(deepcopy(doc), q, a, eng, ii))
        return res

    elif re.search(r"\.(txt)$", filename, re.IGNORECASE):
        callback(0.1, "Start to parse.")
        txt = get_text(filename, binary)
        lines = txt.split("\n")
        comma, tab = 0, 0
        for line in lines:
            if len(line.split(",")) == 2:
                comma += 1
            if len(line.split("\t")) == 2:
                tab += 1
        delimiter = "\t" if tab >= comma else ","

        fails = []
        content = ""
        i = 0
        while i < len(lines):
            arr = lines[i].split(delimiter)
            if len(arr) != 2:
                content += "\n" + lines[i]
            elif len(arr) == 2:
                content += "\n" + arr[0]
                res.append(beAdoc(deepcopy(doc), content, arr[1], eng, i))
                content = ""
            i += 1
            if len(res) % 999 == 0:
                callback(len(res) * 0.6 / len(lines), ("Extract TAG: {}".format(len(res)) + (
                    f"{len(fails)} failure, line: %s..." % (",".join(fails[:3])) if fails else "")))

        callback(0.6, ("Extract TAG: {}".format(len(res)) + (
            f"{len(fails)} failure, line: %s..." % (",".join(fails[:3])) if fails else "")))

        return res

    elif re.search(r"\.(csv)$", filename, re.IGNORECASE):
        callback(0.1, "Start to parse.")
        txt = get_text(filename, binary)
        lines = txt.split("\n")

        fails = []
        content = ""
        res = []
        reader = csv.reader(lines)

        for i, row in enumerate(reader):
            row = [r.strip() for r in row if r.strip()]
            if len(row) != 2:
                content += "\n" + lines[i]
            elif len(row) == 2:
                content += "\n" + row[0]
                res.append(beAdoc(deepcopy(doc), content, row[1], eng, i))
                content = ""
            if len(res) % 999 == 0:
                callback(len(res) * 0.6 / len(lines), ("Extract Tags: {}".format(len(res)) + (
                    f"{len(fails)} failure, line: %s..." % (",".join(fails[:3])) if fails else "")))

        callback(0.6, ("Extract TAG : {}".format(len(res)) + (
            f"{len(fails)} failure, line: %s..." % (",".join(fails[:3])) if fails else "")))
        return res

    raise NotImplementedError(
        "Excel, csv(txt) format files are supported.")


def label_question(question, kbs):
    from api.db.services.knowledgebase_service import KnowledgebaseService
    from graphrag.utils import get_tags_from_cache, set_tags_to_cache
    tags = None
    tag_kb_ids = []
    for kb in kbs:
        if kb.parser_config.get("tag_kb_ids"):
            tag_kb_ids.extend(kb.parser_config["tag_kb_ids"])
    if tag_kb_ids:
        all_tags = get_tags_from_cache(tag_kb_ids)
        if not all_tags:
            all_tags = settings.retriever.all_tags_in_portion(kb.tenant_id, tag_kb_ids)
            set_tags_to_cache(tags=all_tags, kb_ids=tag_kb_ids)
        else:
            all_tags = json.loads(all_tags)
        tag_kbs = KnowledgebaseService.get_by_ids(tag_kb_ids)
        if not tag_kbs:
            return tags
        tags = settings.retriever.tag_query(question,
                                              list(set([kb.tenant_id for kb in tag_kbs])),
                                              tag_kb_ids,
                                              all_tags,
                                              kb.parser_config.get("topn_tags", 3)
                                              )
    return tags


if __name__ == "__main__":
    import sys

    def dummy(prog=None, msg=""):
        pass
    chunk(sys.argv[1], from_page=0, to_page=10, callback=dummy)

File: internal/ingestion/component/chunker/tag.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

// TagChunker ports the Python `tag` chunk method (rag/app/tag.py).
//
// The Python chunk() reads an Excel/csv/txt file with two logical
// columns — content and tags (no header) — and emits ONE chunk per
// row, each carrying:
//
//	content_with_weight = content          (verbatim)
//	tag_kwd             = [t for t in tags.split(",") if t]  (".\" -> "_")
//
// The Go ingestion pipeline feeds the chunker a parsed payload rather
// than the raw file, so TagChunker mirrors the sibling QAChunker and
// extracts (content, tags) pairs from every accepted upstream format:
//
//   - Text / Markdown / CSV-style txt → delimiter-based pairing
//     (tab wins over comma, matching tag.py's detectDelimiter).
//   - HTML (spreadsheet → table)         → first two <td> columns.
//   - JSON (parsed doc items)            → per-item text pairing.
//
// Every pair becomes one chunk with content_with_weight = content and
// tag_kwd = the comma-split, dot-sanitised tag list.
package chunker

import (
	"context"
	"fmt"
	"html"
	"strings"

	"ragflow/internal/agent/runtime"
	"ragflow/internal/ingestion/component/schema"
	"ragflow/internal/tokenizer"
)

const ComponentNameTagChunker = "TagChunker"

type tagChunkerParam struct{}

func (p *tagChunkerParam) Update(conf map[string]any) {}

func (tagChunkerParam) Defaults() tagChunkerParam { return tagChunkerParam{} }

func (tagChunkerParam) Validate() error { return nil }

type TagChunkerComponent struct {
	name  string
	param tagChunkerParam
}

func NewTagChunker(params map[string]any) (runtime.Component, error) {
	p := tagChunkerParam{}.Defaults()
	(&p).Update(params)
	if err := p.Validate(); err != nil {
		return nil, err
	}
	return &TagChunkerComponent{
		name:  ComponentNameTagChunker,
		param: p,
	}, nil
}
func (c *TagChunkerComponent) Inputs() map[string]string { return ChunkerInputs }

func (c *TagChunkerComponent) Outputs() map[string]string { return ChunkerOutputs }

func (c *TagChunkerComponent) Invoke(ctx context.Context, inputs map[string]any) (map[string]any, error) {
	return c.invoke(ctx, inputs)
}

func (c *TagChunkerComponent) invoke(_ context.Context, inputs map[string]any) (map[string]any, error) {
	if inputs == nil {
		return emptyOutputs(), nil
	}
	upstream, err := decodeChunkerFromUpstream(inputs)
	if err != nil {
		return map[string]any{
			"output_format": "chunks",
			"chunks":        []map[string]any{},
			"_ERROR":        fmt.Sprintf("Input error: %v", err),
		}, nil
	}

	var pairs []tagPair
	switch upstream.OutputFormat {
	case schema.PayloadFormatHTML:
		pairs = extractTagTable(stringPtrVal(upstream.HTMLResult))
	case schema.PayloadFormatMarkdown:
		pairs = extractTagText(stringPtrVal(upstream.MarkdownResult))
	case schema.PayloadFormatText:
		pairs = extractTagText(stringPtrVal(upstream.TextResult))
	default:
		pairs = extractTagJSON(upstream.JSONResult)
	}

	chunks := make([]schema.ChunkDoc, 0, len(pairs))
	for _, pair := range pairs {
		content := strings.TrimSpace(pair.Content)
		if content == "" {
			continue
		}
		contentLTKS, _ := tokenizer.Tokenize(content)
		contentSMLTKS, _ := tokenizer.FineGrainedTokenize(contentLTKS)
		chunk := schema.ChunkDoc{
			ContentWithWeight: content,
			DocType:           "text",
			ContentLtks:       contentLTKS,
			ContentSmLtks:     contentSMLTKS,
			TagKwd:            splitTagKwd(pair.Tags),
		}
		chunks = append(chunks, chunk)
	}

	return chunkOutputs(chunks), nil
}

// tagPair is a (content, tags) row extracted from the upstream payload.
type tagPair struct {
	Content string
	Tags    string
}

// extractTagText ports tag.py:60-89 (txt) and tag.py:91-113 (csv):
// every deformed line is accumulated into the running content; a line that
// splits into exactly two fields is emitted as a (content+prefix, tags)
// pair and the accumulator is reset. Delimiter selection (tab vs comma)
// reuses the same detectDelimiter/splitQA helpers as QAChunker.
func extractTagText(text string) []tagPair {
	if text == "" {
		return nil
	}
	lines := strings.Split(text, "\n")
	delimiter := detectDelimiter(lines)

	var pairs []tagPair
	content := ""
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		parts := splitQA(line, delimiter)
		if len(parts) != 2 {
			content += "\n" + line
			continue
		}
		content += "\n" + parts[0]
		pairs = append(pairs, tagPair{Content: content, Tags: parts[1]})
		content = ""
	}
	return pairs
}

// extractTagTable ports the spreadsheet (xlsx/csv → html table) path.
// Python tag.py reads these via the Excel parser's (q, a) pairs, which
// are exactly the first two columns of each row — so we take the same
// two leading cells per <tr>.
func extractTagTable(htmlStr string) []tagPair {
	if htmlStr == "" {
		return nil
	}
	rows := htmlTR.FindAllStringSubmatch(htmlStr, -1)
	pairs := make([]tagPair, 0, len(rows))
	for _, row := range rows {
		cells := htmlTD.FindAllStringSubmatch(row[1], -1)
		var texts []string
		for _, cell := range cells {
			t := html.UnescapeString(htmlTag.ReplaceAllString(cell[1], ""))
			t = strings.TrimSpace(t)
			if t != "" {
				texts = append(texts, t)
			}
		}
		if len(texts) >= 2 {
			pairs = append(pairs, tagPair{Content: texts[0], Tags: texts[1]})
		}
	}
	return pairs
}

// extractTagJSON ports the JSON/structured upstream path: each item's
// text is run through the same delimiter pairing as the text path.
func extractTagJSON(items []schema.ChunkDoc) []tagPair {
	var pairs []tagPair
	for _, item := range items {
		txt, _ := itemText(item)
		if txt == "" {
			continue
		}
		pairs = append(pairs, extractTagText(txt)...)
	}
	return pairs
}

// splitTagKwd ports tag.py:beAdoc's tag_kwd construction: split on
// comma, drop empties, and replace "." with "_" (the storage layer
// rejects dots in keyword values).
func splitTagKwd(s string) []string {
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		p = strings.ReplaceAll(p, ".", "_")
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

func init() {
	MustRegisterChunker(ComponentNameTagChunker)
}
